ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V2/LSM.hs

```diff
@@ -0,0 +1,420 @@
+  ResourceRegistry m ->
+  Session m ->
+  LedgerTables l ValuesMK ->
+  m (ResourceKey m, UTxOTable m l)
+tableFromValuesMK rr session (LedgerTables (ValuesMK values)) = do
+  res@(_, table) <- allocate rr (\_ -> LSM.newTable session) LSM.closeTable
+  LSM.inserts table $
+    V.fromList $
+      [(k, toLSMTxOut (Proxy @l) v, Nothing) | (k, v) <- Map.toList values]
+  pure res
+
+guardClosed ::
+  MonadSTM m => StrictTVar m (LedgerTablesHandleState m l) -> (UTxOTable m l -> m a) -> m a
+guardClosed tv f =
+  readTVarIO tv >>= \case
+    LedgerTablesHandleClosed -> error $ show LSMClosedExn
+    LedgerTablesHandleOpen st -> f st
+
+snapshotManager ::
+  ( IOLike m
+  , LedgerDbSerialiseConstraints blk
+  , LedgerSupportsProtocol blk
+  ) =>
+  Session m ->
+  Complete LedgerDbArgs m blk ->
+  SnapshotManager m m blk (StateRef m (ExtLedgerState blk))
+snapshotManager session args =
+  snapshotManager'
+    session
+    (configCodec . getExtLedgerCfg . ledgerDbCfg $ lgrConfig args)
+    (LedgerDBSnapshotEvent >$< lgrTracer args)
+    (lgrHasFS args)
+
+snapshotManager' ::
+  ( IOLike m
+  , LedgerDbSerialiseConstraints blk
+  , LedgerSupportsProtocol blk
+  ) =>
+  Session m ->
+  CodecConfig blk ->
+  Tracer m (TraceSnapshotEvent blk) ->
+  SomeHasFS m ->
+  SnapshotManager m m blk (StateRef m (ExtLedgerState blk))
+snapshotManager' session ccfg tracer fs =
+  SnapshotManager
+    { listSnapshots = defaultListSnapshots fs
+    , deleteSnapshot = implDeleteSnapshot session fs tracer
+    , takeSnapshot = implTakeSnapshot ccfg tracer fs
+    }
+
+newLSMLedgerTablesHandle ::
+  forall m l.
+  ( IOLike m
+  , HasLedgerTables l
+  , LedgerSupportsLSMLedgerDB l
+  ) =>
+  Tracer m V2.FlavorImplSpecificTrace ->
+  ResourceRegistry m ->
+  Session m ->
+  (ResourceKey m, UTxOTable m l) ->
+  m (LedgerTablesHandle m l)
+newLSMLedgerTablesHandle tracer rr session (resKey, t0) = do
+  !tv <- newTVarIO (LedgerTablesHandleOpen t0)
+  traceWith tracer V2.TraceLedgerTablesHandleCreate
+  pure
+    LedgerTablesHandle
+      { close = do
+          Monad.void $ release resKey
+          atomically $ writeTVar tv LedgerTablesHandleClosed
+          traceWith tracer V2.TraceLedgerTablesHandleClose
+      , duplicate = guardClosed tv $ \t -> do
+          table <- allocate rr (\_ -> LSM.duplicate t) LSM.closeTable
+          newLSMLedgerTablesHandle tracer rr session table
+      , read = \st (LedgerTables (KeysMK keys)) ->
+          guardClosed
+            tv
+            ( \t -> do
+                let keys' = Set.toList keys
+                res <- LSM.lookups t (V.fromList keys')
+                pure $
+                  LedgerTables $
+                    ValuesMK $
+                      Map.fromList [(k, fromLSMTxOut st v) | (k, LSM.Found v) <- zip keys' (V.toList res)]
+            )
+      , readRange = implReadRange tv
+      , readAll = \st ->
+          let readAll' m = do
+                (v, n) <- implReadRange tv st (m, 100000)
```

Comment:
Is there any particular reason why we chose this constant? Would it make sense to add it as a parameter, and pass it in some top-level function with a corresponding note?

Reply:
Quick back of the envelope calculation, if we assume keys are ~34 bytes and values are ~60 bytes. `1_000_000` of these key-value pairs would amount to ~100 MB of disk data.

```
ghci> (1000000*(34+60))/1024/1024
89.6453857421875
```

Seeing as `readAll` reads all the disk data into memory anyway, which is inherently memory and CPU intensive, maybe the read size parameter doesn't matter so much. Still, could be good to test out different values

Reply:
readAll should be gone once we have streaming tables, so I wouldn't worry too much, but I will leave this comment open just in case